File: fabric-model-readiness/src/scout/scorer.py
"""Computes the 0-100 readiness score from a list of findings."""

from __future__ import annotations

from shared.config import CATEGORY_WEIGHTS, CRITICAL_PENALTY_MULTIPLIER
from shared.model import Finding, ScanSummary, Severity


def compute_summary(findings: list[Finding], total_checks_by_category: dict[str, int]) -> ScanSummary:
    """Compute severity counts and weighted readiness score.

    Args:
        findings: All findings from the rule engine.
        total_checks_by_category: Total checks run per category key
            (e.g. {"schema_design": 15, "metadata_completeness": 42, ...}).
            If a category had zero checks, it is excluded from scoring.
    """
    summary = ScanSummary()

    # Count by severity
    for f in findings:
        match f.severity:
            case Severity.CRITICAL:
                summary.critical += 1
            case Severity.HIGH:
                summary.high += 1
            case Severity.MEDIUM:
                summary.medium += 1
            case Severity.LOW:
                summary.low += 1
            case Severity.INFO:
                summary.info += 1

    # Weighted score
    failures_by_cat: dict[str, int] = {}
    for f in findings:
        cat = f.category.value
        penalty = CRITICAL_PENALTY_MULTIPLIER if f.severity == Severity.CRITICAL else 1
        failures_by_cat[cat] = failures_by_cat.get(cat, 0) + penalty

    weighted_score = 0.0
    total_weight = 0.0
    for cat, weight in CATEGORY_WEIGHTS.items():
        total = total_checks_by_category.get(cat, 0)
        if total == 0:
            continue
        failed = failures_by_cat.get(cat, 0)
        passing = max(0, total - failed)
        cat_score = passing / total
        weighted_score += cat_score * weight
        total_weight += weight

    if total_weight > 0:
        summary.score = round((weighted_score / total_weight) * 100, 1)
    else:
        summary.score = 0.0

    return summary


def rating(score: float) -> str:
    """Return a human-readable rating string."""
    if score >= 90:
        return "AI-Ready"
    elif score >= 75:
        return "Mostly Ready (minor gaps)"
    elif score >= 50:
        return "Needs Work (significant gaps)"
    else:
        return "Not Ready (fundamental issues)"

File: docs/src/scanner/scorer.ts
import type { Category, Finding, ScanSummary } from "./types";

const CATEGORY_WEIGHTS: Record<Category, number> = {
  ai_preparation: 0.20,
  metadata_completeness: 0.20,
  schema_design: 0.15,
  measures: 0.15,
  relationships: 0.10,
  data_types: 0.05,
  data_consistency: 0.05,
  org_standards: 0.10,
};

const CRITICAL_PENALTY_MULTIPLIER = 2;

export function computeSummary(
  findings: Finding[],
  totalChecksByCategory: Record<string, number>,
): ScanSummary {
  const summary: ScanSummary = { critical: 0, high: 0, medium: 0, low: 0, info: 0, score: 0 };

  // Count by severity
  for (const f of findings) {
    summary[f.severity] += 1;
  }

  // Weighted score
  const failuresByCat: Record<string, number> = {};
  for (const f of findings) {
    const penalty = f.severity === "critical" ? CRITICAL_PENALTY_MULTIPLIER : 1;
    failuresByCat[f.category] = (failuresByCat[f.category] ?? 0) + penalty;
  }

  let weightedScore = 0;
  let totalWeight = 0;

  for (const [cat, weight] of Object.entries(CATEGORY_WEIGHTS)) {
    const total = totalChecksByCategory[cat] ?? 0;
    if (total === 0) continue;
    const failed = failuresByCat[cat] ?? 0;
    const passing = Math.max(0, total - failed);
    const catScore = passing / total;
    weightedScore += catScore * weight;
    totalWeight += weight;
  }

  summary.score = totalWeight > 0 ? Math.round((weightedScore / totalWeight) * 1000) / 10 : 0;

  return summary;
}

export function rating(score: number): string {
  if (score >= 90) return "AI-Ready";
  if (score >= 75) return "Mostly Ready (minor gaps)";
  if (score >= 50) return "Needs Work (significant gaps)";
  return "Not Ready (fundamental issues)";
}

/**
 * Estimate total checks per category from findings.
 * Uses the heuristic that findings represent ~40% of total checks (pass rate ~60%).
 * This matches the Python CLI behavior.
 */
export function estimateTotalChecks(findings: Finding[]): Record<string, number> {
  const countByCat: Record<string, number> = {};
  for (const f of findings) {
    countByCat[f.category] = (countByCat[f.category] ?? 0) + 1;
  }

  const totals: Record<string, number> = {};
  for (const [cat, count] of Object.entries(countByCat)) {
    totals[cat] = Math.max(count, Math.ceil(count / 0.6));
  }

  // Ensure all categories are represented with at least 1 check
  for (const cat of Object.keys(CATEGORY_WEIGHTS)) {
    if (!(cat in totals)) {
      totals[cat] = 1;
    }
  }

  return totals;
}
